### PR TITLE
[FLINK-30798][runtime] Make OutputFormat support speculative execution for batch jobs

### PR DESCRIPTION
**Pull request by ifndef-SleePy**
## What is the purpose of the change

*This pull request introduces a way to enable speculative execution for `OutputFormat`*

## Brief change log

  - *Introduce context in `OutputFormat` and `FinalizeOnMaster` to expose necessary infos for speculative execution*
  - *Speculative execution would be enabled if a `OutputFormat` inherits SupportsConcurrentExecutionAttempts*

## Verifying this change

This change added tests:

  - *Add new test cases in `StreamingJobGraphGeneratorTest`*
  - *Add new test cases in `SpeculativeSchedulerITCase`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by ifndef-SleePy** [comment]
@zhuzhurk Please take a look when you have time.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java [hunk 1]
```diff
@@ -579,6 +579,28 @@ public interface InitializeOnMasterContext {
         int getExecutionParallelism();
     }
 
+    /** The context exposes some runtime infos for finalization. */
+    public interface FinalizationOnMasterContext {
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java:583]: FinalizationOnMasterContext -> FinalizeOnMasterContext

To be aligned with `InitializeOnMasterContext`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java [hunk 2]
```diff
@@ -579,6 +579,28 @@ public interface InitializeOnMasterContext {
...
+         * might be updated e.g. by the {@link
+         * org.apache.flink.runtime.scheduler.adaptive.AdaptiveScheduler}.
+         */
+        int getExecutionParallelism();
+
+        /**
+         * Get the finished attempt number of subtask.
+         *
+         * @param subtaskIndex the subtask index.
+         * @return the finished attempts.
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java:599]: attempts -> attempt

**Review by zhuzhurk — COMMENTED** [review]
Thanks for creating the PR! @ifndef-SleePy 
The change looks quite good to me except for 2 minor comments.

File: flink-core/src/main/java/org/apache/flink/api/common/io/FinalizeOnMaster.java [hunk 3]
```diff
@@ -66,6 +66,7 @@ interface FinalizationContext {
          *
          * @param subtaskIndex the subtask index.
          * @return the finished attempts.
+         * @throws IllegalArgumentException Thrown, if subtaskIndex is invalid.
```
Comment (zhuzhurk) [inline flink-core/src/main/java/org/apache/flink/api/common/io/FinalizeOnMaster.java:69]: This change should be part of a previous commit. This also applies some other change in this commit.

Comment (ifndef-SleePy) [reply flink-core/src/main/java/org/apache/flink/api/common/io/FinalizeOnMaster.java:69]: Oops, you are right. I squash the wrong commit.

**Comment by ifndef-SleePy** [comment]
@zhuzhurk Thanks for the detailed review! I've addressed the comments. Please take a look.

**Review by zhuzhurk — APPROVED** [review]
Thanks for addressing all the comments! @ifndef-SleePy 
LGTM.

**Comment by zhuzhurk** [comment]
@flinkbot run azure

**Comment by zhuzhurk** [comment]
Merging.

**Pull request closed without merging**